lib/web_ui/lib/src/engine/maybe_shadow_root.dart

```diff
@@ -0,0 +1,144 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+part of engine;
+
+/// This interface defines the common functionality of [html.ShadowRoot] and [html.Document].
+///
+/// Do not confuse this with [html.DocumentOrShadowRoot].
+abstract class ShadowRootOrDocument {
```

Comment:
I think instead of the conditional logic in the implementation I'd prefer an interface and two concrete implementations, one for shadow DOM and one for a plain element. Something like:

```dart
abstract class HostNode {
}

class ShadowDomHostNode implements HostNode {
}

class ElementHostNode implements HostNode {
}
```

The decision to pick between `ShadowDomHostNode` and `ElementHostNode` can be made once and in a single location in the code.

Would that work? Or am I missing something that makes this not viable?

Reply:
Nope, I thought about this too, and I prefer it. The only concern I have is that the core logic for this becomes part of the dom_renderer file, which is quite harder to unit test properly :P

Reply:
What core logic would have to be in the dom_renderer?

Reply:
> What core logic would have to be in the dom_renderer?

Just the try/catch to create either instance.